kernel/cbytegen.ml

```diff
@@ -829,38 +683,53 @@ let rec compile_constr reloc c sz cont =
         | Kbranch lbl -> Kpush_retaddr lbl ::  !c
 	| _ -> !c
       in
-      compile_constr reloc a sz
-      (try
-	let entry = mkInd ind in
-	Retroknowledge.get_vm_before_match_info (!global_env).retroknowledge
-	                                       entry code_sw
-      with Not_found ->
-	code_sw)
-
-and compile_str_cst reloc sc sz cont =
-  set_max_stack_size sz;
-  match sc with
-  | Bconstr c -> compile_constr reloc c sz cont
-  | Bstrconst sc -> Kconst sc :: cont
-  | Bmakeblock(tag,args) ->
-      let arity = Array.length args in
-      let cont = code_makeblock ~stack_size:(sz+arity-1) ~arity ~tag cont in
-      comp_args compile_str_cst reloc args sz cont
-  | Bconstruct_app(tag,nparams,arity,args) ->
-      if Int.equal (Array.length args) 0 then
-        code_construct tag nparams arity cont
-      else
-	comp_app
-	  (fun _ _ _ cont -> code_construct tag nparams arity cont)
-	  compile_str_cst reloc () args sz cont
-  | Bspecial (comp_fx, args) -> comp_fx reloc args sz cont
+      compile_lam env reloc a sz code_sw
+
+  | Lmakeblock (tag,args) ->
+    let arity = Array.length args in
+    let cont = code_makeblock ~stack_size:(sz+arity-1) ~arity ~tag cont in
+    comp_args (compile_lam env) reloc args sz cont
+
+  | Lprim (kn, ar, op, args) ->
+    op_compilation env ar op kn reloc args sz cont
+
+  | Luint v ->
+     (match v with
+     | UintVal i -> Kconst (Const_b0 (Uint31.to_int i)) :: cont
+     | UintDigits ds ->
+       let nargs = Array.length ds in
+       if Int.equal nargs 31 then
+         let (escape,labeled_cont) = make_branch cont in
+         let else_lbl = Label.create() in
+         comp_args (compile_lam env) reloc ds sz
+           ( Kisconst else_lbl::Kareconst(30,else_lbl)::Kcompint31::escape::Klabel else_lbl::Kmakeblock(31, 1)::labeled_cont)
+       else
+         let code_construct cont = (* spiwack: variant of the global code_construct
+                                             which handles dynamic compilation of
+                                             integers *)
+           let f_cont =
+             let else_lbl = Label.create () in
+             [Kacc 0; Kpop 1; Kisconst else_lbl; Kareconst(30,else_lbl);
+              Kcompint31; Kreturn 0; Klabel else_lbl; Kmakeblock(31, 1); Kreturn 0]
+           in
+           let lbl = Label.create() in
+           fun_code := [Ksequence (add_grab 31 lbl f_cont,!fun_code)];
+           Kclosure(lbl,0) :: cont
+         in
+         if Int.equal nargs 0 then
+           code_construct cont
```

Comment:
Given that you have added this special case to `comp_app` already, this branch seems redundant.